File: bot/db_init.py
import json
import psycopg2
import os
import time
import logging

# Настройки из переменных окружения
DB_HOST = os.getenv("DB_HOST", "localhost")
DB_NAME = os.getenv("DB_NAME", "my_video_data")
DB_USER = os.getenv("DB_USER", "myuser")
DB_PASS = os.getenv("DB_PASS", "mypassword")
JSON_FILE = 'videos.json'

def get_conn():
    while True:
        try:
            conn = psycopg2.connect(
                host=DB_HOST,
                database=DB_NAME,
                user=DB_USER,
                password=DB_PASS
            )
            return conn
        except psycopg2.OperationalError:
            logging.info("Ожидание запуска базы данных...")
            time.sleep(2)

def init_db():
    conn = get_conn()
    cursor = conn.cursor()

    # 1. Создаем таблицы (типы данных адаптированы под Postgres)
    cursor.execute("""
    CREATE TABLE IF NOT EXISTS videos (
        id UUID PRIMARY KEY,
        creator_id TEXT,
        video_created_at TIMESTAMPTZ,
        views_count BIGINT,
        likes_count BIGINT,
        comments_count BIGINT,
        reports_count BIGINT,
        created_at TIMESTAMPTZ,
        updated_at TIMESTAMPTZ
    );
    """)

    cursor.execute("""
    CREATE TABLE IF NOT EXISTS snapshots (
        id UUID PRIMARY KEY,
        video_id UUID REFERENCES videos(id),
        views_count BIGINT,
        likes_count BIGINT,
        comments_count BIGINT,
        reports_count BIGINT,
        delta_views_count BIGINT,
        delta_likes_count BIGINT,
        delta_comments_count BIGINT,
        delta_reports_count BIGINT,
        created_at TIMESTAMPTZ,
        updated_at TIMESTAMPTZ
    );
    """)
    conn.commit()

    # 2. Читаем JSON
    try:
        with open(JSON_FILE, 'r', encoding='utf-8') as f:
            data = json.load(f)
    except FileNotFoundError:
        logging.error("videos.json не найден!")
        return

    videos_list = data.get("videos", [])
    logging.info(f"Начинаем импорт {len(videos_list)} видео в PostgreSQL...")

    # 3. Импорт данных
    try:
        for video in videos_list:
            cursor.execute("""
                INSERT INTO videos (id, creator_id, video_created_at, views_count, likes_count, comments_count, reports_count, created_at, updated_at)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
                ON CONFLICT (id) DO UPDATE SET
                views_count = EXCLUDED.views_count,
                likes_count = EXCLUDED.likes_count,
                updated_at = EXCLUDED.updated_at;
            """, (
                video.get('id'), video.get('creator_id'), video.get('video_created_at'),
                video.get('views_count'), video.get('likes_count'), video.get('comments_count'),
                video.get('reports_count'), video.get('created_at'), video.get('updated_at')
            ))

            for snap in video.get('snapshots', []):
                cursor.execute("""
                    INSERT INTO snapshots (id, video_id, views_count, likes_count, comments_count, reports_count, 
                     delta_views_count, delta_likes_count, delta_comments_count, delta_reports_count, created_at, updated_at)
                    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                    ON CONFLICT (id) DO NOTHING;
                """, (
                    snap.get('id'), video.get('id'),
                    snap.get('views_count'), snap.get('likes_count'), snap.get('comments_count'), snap.get('reports_count'),
                    snap.get('delta_views_count'), snap.get('delta_likes_count'), snap.get('delta_comments_count'), snap.get('delta_reports_count'),
                    snap.get('created_at'), snap.get('updated_at')
                ))

        conn.commit()
        logging.info("Импорт завершен успешно!")

    except Exception as e:
        logging.error(f"Ошибка при импорте: {e}")
        conn.rollback()
    finally:
        cursor.close()
        conn.close()

if __name__ == "__main__":
    init_db()

File: db_init.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <libpq-fe.h>
#include <cjson/cJSON.h>

char* read_file(const char* filename) {
    FILE *f = fopen(filename, "rb");
    if (!f) return NULL;
    fseek(f, 0, SEEK_END);
    long length = ftell(f);
    fseek(f, 0, SEEK_SET);
    char *data = malloc(length + 1);
    fread(data, 1, length, f);
    data[length] = '\0';
    fclose(f);
    return data;
}

int main() {
    const char *db_host = getenv("DB_HOST") ? getenv("DB_HOST") : "localhost";
    const char *db_user = getenv("DB_USER") ? getenv("DB_USER") : "myuser";
    const char *db_pass = getenv("DB_PASS") ? getenv("DB_PASS") : "mypassword";
    const char *db_name = getenv("DB_NAME") ? getenv("DB_NAME") : "my_video_data";

    char conninfo[256];
    snprintf(conninfo, sizeof(conninfo), "host=%s user=%s password=%s dbname=%s", db_host, db_user, db_pass, db_name);

    PGconn *conn = PQconnectdb(conninfo);
    if (PQstatus(conn) != CONNECTION_OK) {
        fprintf(stderr, "DB Error: %s\n", PQerrorMessage(conn));
        PQfinish(conn);
        return 1;
    }

    const char *sql_videos = "CREATE TABLE IF NOT EXISTS videos (id UUID PRIMARY KEY, creator_id TEXT, video_created_at TIMESTAMPTZ, views_count BIGINT, likes_count BIGINT, comments_count BIGINT, reports_count BIGINT, created_at TIMESTAMPTZ, updated_at TIMESTAMPTZ);";
    const char *sql_snapshots = "CREATE TABLE IF NOT EXISTS snapshots (id UUID PRIMARY KEY, video_id UUID REFERENCES videos(id), views_count BIGINT, likes_count BIGINT, comments_count BIGINT, reports_count BIGINT, delta_views_count BIGINT, delta_likes_count BIGINT, delta_comments_count BIGINT, delta_reports_count BIGINT, created_at TIMESTAMPTZ, updated_at TIMESTAMPTZ);";

    PQclear(PQexec(conn, sql_videos));
    PQclear(PQexec(conn, sql_snapshots));

    printf("Читаем videos.json...\n");
    char *json_data = read_file("videos.json");
    if (!json_data) { printf("Файл videos.json не найден!\n"); return 1; }

    cJSON *root = cJSON_Parse(json_data);
    cJSON *videos = cJSON_GetObjectItem(root, "videos");
    cJSON *video;

    cJSON_ArrayForEach(video, videos) {
        const char *v_params[9] = {
            cJSON_GetObjectItem(video, "id")->valuestring,
            cJSON_GetObjectItem(video, "creator_id")->valuestring,
            cJSON_GetObjectItem(video, "video_created_at")->valuestring,
            cJSON_PrintUnformatted(cJSON_GetObjectItem(video, "views_count")),
            cJSON_PrintUnformatted(cJSON_GetObjectItem(video, "likes_count")),
            cJSON_PrintUnformatted(cJSON_GetObjectItem(video, "comments_count")),
            cJSON_PrintUnformatted(cJSON_GetObjectItem(video, "reports_count")),
            cJSON_GetObjectItem(video, "created_at")->valuestring,
            cJSON_GetObjectItem(video, "updated_at")->valuestring
        };

        const char *insert_v = "INSERT INTO videos VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9) ON CONFLICT (id) DO UPDATE SET views_count=EXCLUDED.views_count, likes_count=EXCLUDED.likes_count, updated_at=EXCLUDED.updated_at;";
        PQclear(PQexecParams(conn, insert_v, 9, NULL, v_params, NULL, NULL, 0));

        for(int i=3; i<=6; i++) free((void*)v_params[i]); // Очищаем память от чисел

        cJSON *snapshots = cJSON_GetObjectItem(video, "snapshots");
        cJSON *snap;
        cJSON_ArrayForEach(snap, snapshots) {
            const char *s_params[12] = {
                cJSON_GetObjectItem(snap, "id")->valuestring,
                cJSON_GetObjectItem(video, "id")->valuestring,
                cJSON_PrintUnformatted(cJSON_GetObjectItem(snap, "views_count")),
                cJSON_PrintUnformatted(cJSON_GetObjectItem(snap, "likes_count")),
                cJSON_PrintUnformatted(cJSON_GetObjectItem(snap, "comments_count")),
                cJSON_PrintUnformatted(cJSON_GetObjectItem(snap, "reports_count")),
                cJSON_PrintUnformatted(cJSON_GetObjectItem(snap, "delta_views_count")),
                cJSON_PrintUnformatted(cJSON_GetObjectItem(snap, "delta_likes_count")),
                cJSON_PrintUnformatted(cJSON_GetObjectItem(snap, "delta_comments_count")),
                cJSON_PrintUnformatted(cJSON_GetObjectItem(snap, "delta_reports_count")),
                cJSON_GetObjectItem(snap, "created_at")->valuestring,
                cJSON_GetObjectItem(snap, "updated_at")->valuestring
            };

            const char *insert_s = "INSERT INTO snapshots VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12) ON CONFLICT (id) DO NOTHING;";
            PQclear(PQexecParams(conn, insert_s, 12, NULL, s_params, NULL, NULL, 0));
            for(int i=2; i<=9; i++) free((void*)s_params[i]);
        }
    }

    cJSON_Delete(root);
    free(json_data);
    PQfinish(conn);
    printf("Импорт завершен успешно!\n");
    return 0;
}